### PR TITLE
[lexical-playground][lexical] Feature: Ruby annotation node with floating editor

### LINKED ISSUE #7787: Feature: support of ruby characters
## Description

I'm working on a rich text editor for east asian users. I would like to have the support of ruby characters.

Ruby characters (also known as ruby annotations or furigana) are small, annotative characters placed above or beside base characters in East Asian texts. They are commonly used to provide pronunciation guides (e.g., hiragana/katakana for kanji in Japanese, pinyin for Chinese characters) or semantic explanations.

Currently, Lexical does not natively support the creation or editing of ruby characters, which limits its usability for East Asian languages. Implementing this feature would enable developers to build rich text editors that fully support Japanese, Chinese, and other East Asian languages with ruby annotations.

```html
<ruby>
  <rb>振</rb><rb>り</rb><rb>仮</rb><rb>名</rb>
  <rp>(</rp><rt>ふ</rt><rt>り</rt><rt>が</rt><rt>な</rt><rp>)</rp>
</ruby>
```

Reference -> 
https://www.w3.org/TR/css-ruby-1/#:~:text=%3Cruby%3E%0A%20%20%3Crb%3E%E6%8C%AF%3C/rb%3E%3Crb%3E%E3%82%8A%3C/rb%3E%3Crb%3E%E4%BB%AE%3C/rb%3E%3Crb%3E%E5%90%8D%3C/rb%3E%0A%20%20%3Crp%3E(%3C/rp%3E%3Crt%3E%E3%81%B5%3C/rt%3E%3Crt%3E%E3%82%8A%3C/rt%3E%3Crt%3E%E3%81%8C%3C/rt%3E%3Crt%3E%E3%81%AA%3C/rt%3E%3Crp%3E)%3C/rp%3E%0A%3C/ruby%3E

## Impact

Enterprise Software Companies: Building applications for East Asian markets (e.g., Slack, Notion, Microsoft Office alternatives)

Developers: Creating specialized content management systems for academic, literary, or linguistic applications

### PR DESCRIPTION
**Pull request by mayrang**
## Description

Adds a playground-level `RubyNode` (extends `TextNode` in token mode) with full keyboard navigation, IME composition support, and a floating annotation editor. The node renders as `<ruby><rt>` on export and imports the same structure back.

### What changed

**RubyNode** (`plugins/RubyExtension/RubyNode.ts`) — a token-mode `TextNode` subclass with an `annotation` state (via `createState`). The DOM is a wrapper `<span>` containing an inner `<span data-ruby-annotation="...">` so the annotation text renders via CSS `::before`. `exportDOM` produces standard `<ruby>...<rt>...</rt></ruby>`.

**RubyExtension** (`plugins/RubyExtension/index.ts`) — registers arrow-key, backspace, and composition handlers. Arrow keys skip over consecutive ruby groups atomically — ruby base text is in token mode, so the cursor can't meaningfully stop on it (any keystroke would be redirected to a neighbor). Skipping the entire contiguous group avoids a confusing dead-stop. Shift+arrow extends selection past them. A `SELECTION_CHANGE` handler nudges collapsed cursors off non-composing ruby nodes to prevent the caret from landing inside a token.

**Floating editor** (`plugins/RubyExtension/FloatingRubyEditor.tsx`) — follows the `FloatingLinkEditorPlugin` pattern. The editor appears anchored to the selection when creating new ruby (toolbar trigger) or to the ruby DOM element when clicking an existing one. Enter confirms, Escape dismisses. While active, the `FloatingTextFormatToolbar` is suppressed so the two don't overlap.

**Core changes** (`LexicalEvents.ts`, `LexicalUtils.ts`) — targeted fixes for IME composition on token-mode `TextNode` subclasses:

1. `$onCompositionEndImpl` now detects when composition ends on a token node and redirects the composed text to the adjacent `TextNode` via `selection.insertText`, instead of letting `markDirty` silently discard it. Uses the actual `selection.anchor.offset` to decide direction — offset 0 redirects to the previous sibling, offset=textLen to the next.
2. `$handleInput` skips `$shouldPreventDefaultAndInsertText` during `insertCompositionText` on token nodes, so the browser's native composition UI stays intact until composition ends.
3. `$updateTextNodeFromDOMContent` bails early when a composing token node's DOM is synced, preventing the reconciler from reverting mid-composition input.
4. `$handleCompositionStart` now redirects composition away from token/segmented nodes by inserting `COMPOSITION_START_CHAR`, which triggers the existing `insertText` boundary logic to create an adjacent `TextNode`. This prevents composition from starting inside nodes that can't accept text modifications — the root cause of #6296 (mention entity destruction on CJK input). Combined with (1), this provides two layers: pre-composition redirect at start, and post-composition redirect at end as a fallback.

These core changes are generic — any token-mode or segmented `TextNode` subclass benefits from them, not just `RubyNode`.

### Toolbar

The `ルビ` katakana text button is replaced with an SVG icon (`ruby.svg`) consistent with other toolbar icons. The button only activates when text is selected; clicking with an existing ruby in the selection removes it.

Closes #7787, #6296

## Test plan

- [x] 47 unit tests (`RubyNode.test.ts`) — arrow skip, Shift+arrow selection, consecutive ruby group walk, line boundary fallback, backspace, guard conditions, HTML import rule.
- [x] 23 unit tests (`RubyComposition.test.ts`) — composition end redirect on token nodes, mid-composition DOM stability, composed text placement, edge cases (solo ruby, paragraph-first ruby, offset 0 redirect).
- [x] 22 e2e tests (`Ruby.spec.mjs`) across Chromium/Firefox/WebKit — insert, DOM structure, arrow navigation, backspace/delete, select-all, toggle off, copy-paste, JSON round-trip, exportDOM, Shift+arrow selection, line boundary navigation.
- [x] `tsc --noEmit` / prettier / eslint clean.
- [x] Manual playground (Chrome, Firefox, Safari on macOS) — all scenarios below pass on all three browsers:

<details>
<summary>Manual test scenarios (all three browsers)</summary>

Setup — paste this structure into each test:
```js
document.addEventListener('copy', (e) => { e.preventDefault(); e.clipboardData.setData('text/html', '前<ruby>漢<rt>かん</rt></ruby><ruby>字<rt>じ</rt></ruby>後'); }, {once: true}); document.execCommand('copy');
```

**Arrow navigation**
- `前|漢` → Right → `漢|字` (one move across ruby boundary)
- `字|後` → Right → `後|` (ruby → plain text)
- `|前` → Right → `前|漢` (plain text → ruby)
- `漢|字` → Left → `前|漢` (ruby boundary, backward)
- Shift+Right extends selection past ruby
- Cmd+Right jumps to end of line

**Typing at boundaries**
- `前|漢` (ruby start) → type `あ` → inserted outside ruby, after `前`
- `字|後` (ruby end) → type `あ` → inserted outside ruby, before `後`

**Backspace / Delete**
- `前|漢` → Backspace → `前` deleted
- `字|後` → Delete → `後` deleted
- Select ruby `漢` → Backspace → ruby node removed

**Selection + delete**
- Drag-select `漢字` → Delete → both rubies deleted, `前後` remains
- Drag-select `漢字後` → Delete → ruby + plain text deleted, `前` remains
- Cmd+A → Delete → all deleted, empty paragraph

**Copy / paste**
- Select ruby → copy → paste → ruby structure preserved
- External `<ruby>` HTML paste → converted to RubyNode
- `<rp>` tags in pasted HTML → ignored correctly

**Undo / Redo**
- Create ruby via toolbar → Cmd+Z → ruby unwrapped
- Cmd+Z → Cmd+Shift+Z → ruby restored

**Ruby create / remove**
- Select text → ruby toolbar icon → floating editor → type annotation → Enter → ruby created
- Click existing ruby → floating editor → edit annotation → Enter → annotation updated
- Click existing ruby → trash button → unwrapped to plain text
- No selection → ruby toolbar icon → nothing happens

**Serialization**
- JSON export contains `type: "ruby"`, `annotation: "かん"`, `text: "漢"`
- JSON round-trip (export → parseEditorState → setEditorState) → no visual change
- HTML export produces `<ruby>漢<rt>かん</rt></ruby>`

**Edge cases**
- Paragraph-start ruby: Left at `|漢` → no movement
- Paragraph-end ruby: Right at `字|` → no movement
- Ruby-only paragraph: select → delete → empty paragraph remains
- Three consecutive rubies: each boundary navigable in one arrow press

</details>

### Design notes

**Package placement**: `RubyNode` lives in `lexical-playground`, not in a dedicated `@lexical/ruby` package. The node and extension are structured for extraction if there's interest — `RubyNode` has no playground dependencies, and `RubyExtension` only depends on `@lexical/html` for the import rule.

**Token-mode TextNode vs. inline ElementNode**: The original design was an inline `ElementNode` subclass (like `LinkNode`) with the base text as a child `TextNode`. During implementation this turned out to be impractical for several reasons:

- Cursor enters the element freely, but ruby base text shouldn't be directly editable — changing 漢 to something else while keeping annotation かん doesn't make sense as a default editing model. Token mode gives this atomicity without extra cursor-trapping logic.
- IME composition on a child `TextNode` inside an inline `ElementNode` creates a cascade of issues with Lexical's reconciler and browser selection normalization at element boundaries. The reconciler tries to restore DOM state mid-composition, and Safari in particular normalizes the cursor back into the element across `compositionstart` / `compositionend`. With token mode, composition is handled by the core token-redirect path (which this PR fixes) rather than fighting per-element boundary conditions.
- Selection across multiple ruby elements produces partial-element selections that need extensive normalization. Token mode treats each ruby as an atomic point in the selection, matching how `<ruby>` behaves in HTML — base text and annotation are a single unit.

The tradeoff is that formatting the base text independently (bold only on the kanji) isn't possible without unwrapping, but this matches standard `<ruby>` semantics.

**Composition visual limitation (Safari)**: When IME composition starts immediately after a ruby node, Safari normalizes the cursor into the preceding ruby's inner `<span>` rather than the adjacent `TextNode`. The composed text therefore appears inside the ruby DOM. Since the annotation is rendered via `::after` on the same element, showing both the annotation and the composing text in the same space would overlap. The extension detects this via `compositionstart` / `compositionupdate` listeners and adds a `--composing` CSS class that hides the `::after` annotation until `compositionend`. The result is that the previous ruby's furigana temporarily disappears during the first character's composition, then reappears once composition ends and the text is redirected to its own `TextNode`.

We tried several approaches to avoid this:
- Moving the cursor to the adjacent `TextNode` programmatically before composition: Safari re-normalizes it back into the ruby span, undoing the move.
- Inserting a ZWSP gap `TextNode` between the ruby and the next node to give Safari a landing target: the gap triggers `$normalizeTextNode` which merges it into the adjacent text, and if guarded with `toggleUnmergeable()`, a transform loop removes and re-creates it indefinitely.
- Modifying DOM during `compositionstart`: breaks the IME state entirely, producing garbled input.

Hiding the annotation during composition was the least-bad option — functional correctness is preserved, and the visual disruption lasts only for the first composed character.

### CODE REVIEW

**Review by potatowagon — COMMENTED** [review]
Reviewed by Navi (Tater Thoughts Bobblehead) on behalf of @potatowagon.

**Assessment: Needs fixes before merge (CI failing) — but the feature design is solid.**

## What this PR does
Major new feature: Ruby annotation node with floating editor for the playground. Ruby annotations (`<ruby>漢<rt>かん</rt></ruby>`) are essential for CJK text to show pronunciation guides (furigana/pinyin). Implementation includes:

- **RubyNode** (`nodes/RubyNode.ts`): TextNode subclass in token mode with `$config()` protocol, annotation state via `createState`, dual DOM representation (wrapper span > inner span with `data-ruby-annotation` CSS annotation via `::after`), semantic `exportDOM` producing `<ruby>/<rt>`.
- **RubyExtension** (`plugins/RubyExtension/index.ts`): Comprehensive extension handling arrow key navigation (skips ruby groups atomically), Shift+arrow selection extension, backspace deletion, `$nudgeOffRuby` for selection normalization, Safari IME composition guards, and `DOMImportExtension` for `<ruby>` HTML import.
- **FloatingRubyEditorPlugin**: Floating UI editor (annotation input) triggered via toolbar button or click on existing ruby node.
- **Core Lexical changes** (`LexicalEvents.ts`, `LexicalUtils.ts`): Modifies `$handleInput`, `$handleCompositionStart`, `$onCompositionEndImpl`, and `$updateTextNodeFromDOMContent` to support composition on token nodes — redirecting composed text to adjacent TextNodes instead of mutating the token.

## What I checked
1. **Architecture**: Token-mode TextNode with wrapper DOM and DOMSlot is the right pattern for inline decorators that carry text. The `$config()` protocol usage is correct.
2. **Arrow navigation**: The `$skipRubyOnArrow` logic correctly walks consecutive ruby groups and handles boundary conditions (ruby as first/last/only child → moves to parent element point).
3. **Composition handling**: The Safari IME composition guard (`$nudgeOffRuby` skips when composing, `$updateTextNodeFromDOMContent` early-returns for composing tokens) is critical — prevents data loss during CJK input.
4. **Core changes**: The modifications to `$onCompositionEndImpl` (now returns `boolean` to indicate token-redirect) and `$handleInput` (new `isCompositionOnToken` guard) are well-scoped. The token redirect at composition-end correctly uses `selection.insertText(data)` which handles the token boundary insertion natively.
5. **Test coverage**: Excellent — 764-line e2e spec (20+ tests), 1395-line unit test, 534-line composition test. Covers arrow skip, shift+arrow, backspace, copy/paste, toggle on/off, serialization, exportDOM, consecutive rubies, boundary cases, guard conditions.

## CI Status — ❌ TWO FAILURES

1. **Integrity (lint)**: `no-shadow` error in `RubyExtension/index.ts` line 229 — variable `sel` shadows the `sel` import from `@lexical/html`. Easy fix: rename the local variable (e.g., `domSel` or `nativeSel`).

2. **E2e tests (11 failures)**: Multiple Ruby e2e tests failing with assertion errors — "Ruby DOM has wrapper span with inner annotated span", "Arrow left skips over ruby node", etc. These are the PR's own new tests failing, suggesting the floating editor or DOM structure in the real browser environment differs slightly from what the tests expect. Needs investigation.

## www compatibility
No www compat concerns — all playground-only changes except the core `LexicalEvents.ts` / `LexicalUtils.ts` modifications. The core changes are additive (new code paths for token+composing) and don't alter existing behavior for non-token nodes. The `$onCompositionEndImpl` return type change from `void` to `boolean` is internal (not exported).

## Suggestions
- Fix the lint `no-shadow` error (rename local `sel` to `nativeSel` on line 229 of RubyExtension/index.ts)
- Investigate the 11 e2e test failures — likely a timing or DOM structure mismatch
- Consider whether the core `LexicalEvents.ts` changes should be in a separate PR for easier review/rollback (they're the highest-risk part)

Overall excellent work — comprehensive feature with thorough test coverage and proper IME handling. Just needs the CI fixes.

**Comment by mayrang** [comment]
IME work is always a handful, but this ruby one was something else 😂

**Review by etrepum — COMMENTED** [review]
Haven't had a chance to take a close look at the code yet but the functionality here is really cool! Seems to work well, was surprised that copy and paste of ruby html from wikipedia worked perfectly.

**Comment by mayrang** [comment]
Thanks — hope it helps users who need ruby annotations!

While auditing open IME/composition issues, I found that #6296 is still reproducible on current main. Added a commit that fixes it.

When IME composition starts on a segmented TextNode (like MentionNode), the browser writes composed text directly into the node's DOM, which triggers the segmented replacement and destroys the entity mid-composition. $handleCompositionStart now redirects composition to a new TextNode via COMPOSITION_START_CHAR — the same mechanism already used for element cursors and format mismatches.

Tested with Chinese pinyin and Japanese IME on Chrome, Firefox, and Safari.

**Comment by potatowagon** [comment]
copied ruby rendered html from https://en.wikipedia.org/wiki/Ruby_character

before:
Screenshot 2026-06-30 at 11 20 19 PM

after:

Screenshot 2026-06-30 at 11 20 11 PM

this is cool, lovely to see the ruby node show up in the lexical tree and dom export

File: packages/lexical-playground/src/nodes/PlaygroundNodes.ts [hunk 1]
```diff
@@ -36,6 +36,7 @@ import {LayoutItemNode} from './LayoutItemNode';
 import {MentionNode} from './MentionNode';
 import {PageBreakNode} from './PageBreakNode';
 import {PollNode} from './PollNode';
+import {RubyNode} from './RubyNode';
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/PlaygroundNodes.ts]: There isn't any reason to update this file for nodes that are configured by extensions

File: packages/lexical-playground/src/__tests__/unit/RubyComposition.test.ts [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+  beforeEach(() => {
+    vi.useFakeTimers();
+    container = document.createElement('div');
+    container.setAttribute('data-lexical-editor', 'true');
+    container.contentEditable = 'true';
+    document.body.appendChild(container);
+    editor = createTestEditor({nodes: [RubyNode]});
+    registerRichText(editor);
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    extensionCleanup = (RubyExtension as any).register(editor);
```
Comment (etrepum) [inline packages/lexical-playground/src/__tests__/unit/RubyComposition.test.ts]: we should really just be using the extension instead of doing weird stuff like this

File: packages/lexical-playground/src/__tests__/unit/RubyNode.test.ts [hunk 3]
```diff
@@ -0,0 +1,1324 @@
...
+  let container: HTMLDivElement;
+  let editor: LexicalEditor;
+
+  beforeEach(() => {
+    container = document.createElement('div');
+    container.setAttribute('data-lexical-editor', 'true');
+    container.contentEditable = 'true';
+    document.body.appendChild(container);
+    editor = createTestEditor({nodes: [RubyNode]});
+    registerRichText(editor);
```
Comment (etrepum) [inline packages/lexical-playground/src/__tests__/unit/RubyNode.test.ts]: these tests should also be using the extension

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  TextNode,
+} from 'lexical';
+
+export type SerializedRubyNode = Spread<
+  {
+    annotation: string;
+  },
+  SerializedTextNode
+>;
+
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]: These serialization types aren't very useful in modern lexical, we don't need to define them. It can be entirely inferred by typescript. Also it's wrong because annotation is actually optional, won't be serialized if it's the default value.

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+/** @noInheritDoc */
+export class RubyNode extends TextNode {
+  $config() {
+    return this.config('ruby', {
+      extends: TextNode,
+      stateConfigs: [{flat: true, stateConfig: annotationState}],
+    });
+  }
+
+  constructor(text: string = '', key?: import('lexical').NodeKey) {
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]: type NodeKey should simply be imported, or you can eliminate the custom constructor entirely by having `$createRubyNode` do `setMode('token')`

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    super(text, key);
+    this.__mode = 1; // token
+  }
+
+  afterCloneFrom(prevNode: this): void {
+    super.afterCloneFrom(prevNode);
+    this.__mode = 1; // token
+  }
+
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]: I think this is entirely redundant, the constructor already sets it and the previous version should not have a conflicting value

```suggestion
```

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      return super.getDOMSlot(dom).withElement(inner);
+    }
+    return super.getDOMSlot(dom);
+  }
+
+  updateDOM(prevNode: this, dom: HTMLElement, config: EditorConfig): boolean {
+    const updated = super.updateDOM(prevNode, dom, config);
+    if (
+      prevNode.getAnnotation() !== this.getAnnotation() ||
+      prevNode.getTextContent() !== this.getTextContent()
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]: This doesn't work because the getters will defer to the latest version, this is what `$getStateChange` and `$getState(node, stateConfig, NODE_STATE_DIRECT)` are for.

textContent would have to be read with `prevNode.__text`

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    const rt = document.createElement('rt');
+    rt.textContent = this.getAnnotation();
+    ruby.appendChild(rt);
+    const rpClose = document.createElement('rp');
+    rpClose.textContent = ')';
+    ruby.appendChild(rpClose);
+    return {element: ruby};
+  }
+
+  getAnnotation(): StateConfigValue<typeof annotationState> {
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]: It can be convenient to have a `version?: NodeStateVersion` argument for node state backed getters

File: packages/lexical-playground/src/nodes/RubyNode.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    return false;
+  }
+
+  canInsertTextAfter(): false {
+    return false;
+  }
+}
+
+export function $createRubyNode(text: string, annotation: string): RubyNode {
+  return new RubyNode(text).setAnnotation(annotation);
```
Comment (etrepum) [inline packages/lexical-playground/src/nodes/RubyNode.ts]:
```suggestion
  return $create(RubyNode).setTextContent(text).setMode('token').setAnnotation(annotation);
```

File: packages/lexical-playground/src/plugins/RubyExtension/RubyNode.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  },
+  SerializedTextNode
+>;
+
+const annotationState = /* @__PURE__ */ createState('annotation', {
+  parse: v => (typeof v === 'string' ? v : ''),
+});
+
+/** @noInheritDoc */
+export class RubyNode extends TextNode {
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/RubyExtension/RubyNode.ts:38]: This node should be colocated with the extension

File: packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,401 @@
+/**
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx:1]: This should probably also be colocated with the extension

File: packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+  event: React.KeyboardEvent<HTMLInputElement> | React.MouseEvent<HTMLElement>,
+): void {
+  event.preventDefault();
+}
+
+function getRubyNodeKeyFromDOM(
+  editor: LexicalEditor,
+  target: EventTarget | null,
+): string | null {
+  if (!(target instanceof HTMLElement)) {
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx]:
```suggestion
  if (!isHTMLElement(target)) {
```

File: packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+        isVisible
+      ) {
+        setIsRubyClick(false);
+        setIsRubyEditMode(false);
+      }
+    };
+    editorElement.addEventListener('focusout', handleBlur);
+    return () => {
+      editorElement.removeEventListener('focusout', handleBlur);
+    };
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx]:
```suggestion
    return registerEventListener(editorElement, 'focusout', handleBlur);
```

File: packages/lexical-playground/src/plugins/RubyExtension/index.ts [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+            'compositionupdate',
+            checkCompositionInRuby,
+            true,
+          );
+          rootElement.addEventListener(
+            'compositionend',
+            onCompositionEnd,
+            true,
+          );
+        }
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/RubyExtension/index.ts]:
```suggestion
        if (rootElement) {
          return registerEventListeners(
            rootElement,
            {
              compositionend: onCompositionEnd,
              compositionstart: checkCompositionInRuby,
              compositionupdate: checkCompositionInRuby,
            },
            true,
          );
        }
```

File: packages/lexical-playground/src/App.tsx [hunk 15]
```diff
@@ -265,6 +266,7 @@ const AppExtension = /* @__PURE__ */ defineExtension({
     DragDropPasteExtension,
     EmojisExtension,
     MentionsExtension,
+    RubyExtension,
```
Comment (etrepum) [inline packages/lexical-playground/src/App.tsx]: This should be in PlaygroundRichTextExtension because it doesn't really work unless it's rich text

File: packages/lexical/src/LexicalEvents.ts [hunk 16]
```diff
@@ -1422,12 +1445,35 @@ function $onCompositionEndImpl(editor: LexicalEditor, data?: string): void {
...
+        const selection = $getSelection();
+        if ($isRangeSelection(selection)) {
+          const textLen = node.getTextContentSize();
+          const offset =
+            selection.anchor.key === compositionKey
+              ? selection.anchor.offset
+              : textLen;
+          selection.anchor.set(compositionKey, offset, 'text');
+          selection.focus.set(compositionKey, offset, 'text');
+          selection.insertText(data);
```
Comment (etrepum) [inline packages/lexical/src/LexicalEvents.ts]:
```suggestion
          node.select(offset, offset).insertText(data);
```

**Review by etrepum — COMMENTED** [review]
Overall this works great, but the tests have some bad practices and the playground could be configured a bit more cleanly

File: packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+      const next = event.relatedTarget as Element | null;
+      if (next !== null) {
+        if (!editorElement.contains(next)) {
+          setIsRubyClick(false);
+          setIsRubyEditMode(false);
+        }
+        return;
+      }
+      requestAnimationFrame(() => {
+        if (editorElement.contains(document.activeElement)) {
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx]: Are these document references correct given the possibility of iframes and shadow roots?

File: packages/lexical-playground/src/nodes/PlaygroundNodes.ts (hunk 1 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/PlaygroundNodes.ts]: Done — removed RubyNode from PlaygroundNodes.

File: packages/lexical-playground/src/__tests__/unit/RubyComposition.test.ts (hunk 2 above)
Comment (mayrang) [reply packages/lexical-playground/src/__tests__/unit/RubyComposition.test.ts]: Done — both test files now use `buildEditorFromExtensions(RubyExtension)`.

File: packages/lexical-playground/src/__tests__/unit/RubyNode.test.ts (hunk 3 above)
Comment (mayrang) [reply packages/lexical-playground/src/__tests__/unit/RubyNode.test.ts]: Done — same as above.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 4 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Done — removed the type entirely.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 5 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Done — eliminated the custom constructor. `$createRubyNode` uses `$create(RubyNode).setMode('token').setAnnotation(annotation)`.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 6 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Done — removed.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 7 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Right — fixed. Using `$getStateChange(this, prevNode, annotationState)` and `prevNode.__text`.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 8 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Done — added `version?: NodeStateVersion` to `getAnnotation`.

File: packages/lexical-playground/src/nodes/RubyNode.ts (hunk 9 above)
Comment (mayrang) [reply packages/lexical-playground/src/nodes/RubyNode.ts]: Done.

File: packages/lexical-playground/src/plugins/RubyExtension/RubyNode.ts (hunk 10 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/RubyExtension/RubyNode.ts:38]: Done — moved to `plugins/RubyExtension/RubyNode.ts`.

File: packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx (hunk 11 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx:1]: Done — moved to `plugins/RubyExtension/FloatingRubyEditor.tsx`.

File: packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx (hunk 12 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx]: Done.

File: packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx (hunk 13 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FloatingRubyEditorPlugin/index.tsx]: Done.

File: packages/lexical-playground/src/plugins/RubyExtension/index.ts (hunk 14 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/RubyExtension/index.ts]: Done.

File: packages/lexical-playground/src/App.tsx (hunk 15 above)
Comment (mayrang) [reply packages/lexical-playground/src/App.tsx]: Done.

File: packages/lexical/src/LexicalEvents.ts (hunk 16 above)
Comment (mayrang) [reply packages/lexical/src/LexicalEvents.ts]: Done.

File: packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx (hunk 17 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/RubyExtension/FloatingRubyEditor.tsx]: No, they weren't — switched all three `document` references to `ownerDocument` for iframe/shadow root compatibility.

**Comment by mayrang** [comment]
All feedback addressed — ran into some issues along the way so it took a bit longer than expected 😅

- Colocated RubyNode and FloatingRubyEditor into the extension folder
- Removed SerializedRubyNode type, custom constructor, and redundant `$transform`
- Fixed `updateDOM` to use `$getStateChange` + `prevNode.__text`
- Added `version` parameter to `getAnnotation`
- Converted both test files to `buildEditorFromExtensions` pattern
- Applied all utility suggestions (`isHTMLElement`, `registerEventListener`, `registerEventListeners`, `node.select()`)
- Moved RubyExtension to PlaygroundRichTextExtension
- Fixed `document` references to use `ownerDocument` for iframe/shadow root compatibility

**Comment by etrepum** [comment]
@mayrang I worked with claude to simplify some of the code to use the caret APIs and add a little more test coverage, let me know if you think something else should be addressed before merge

<details>
<summary>
Suggested changes: `claude/ruby-extension-nodecaret-h1k1if` (1 commit on top of `0a8698d`)
</summary>

### RubyExtension: rewrite arrow navigation on NodeCaret APIs
- `$walkPastRubyChain`'s direction lambdas → `$getSiblingCaret`/`getNodeAtCaret` over a `CaretDirection`
- `$skipRubyOnArrow`'s duplicated edge-detection/destination branches collapse to one path:
  `$caretFromPoint` + `$isExtendableTextPointCaret` for detection, and a single
  `$setPointFromCaret(point, edgeCaret.getFlipped())` for the landing (same boundary,
  materialized off the token node — text points on the ruby would re-enter the handler
  and route IME through the token path; ~21 existing unit tests pin the representation)
- Element points now handled uniformly (previously bailed on `point.type !== 'text'`);
  backspace uses the same primitives and works at the parent-boundary element point
  that arrow navigation itself creates; left/right registrations deduplicated
- The `offset >= 1` shift-landing is kept, comment corrected: it is **not** Safari-specific —
  a focus at offset 0 after the ruby is resolved back onto the ruby end by the DOM
  selection round-trip (reproduced in Chromium), stalling repeated Shift+Right

### FloatingRubyEditor: shadow-safe focus checks
- `ownerDocument.activeElement` fixes iframes but reports the shadow *host* inside a
  shadow root, so the focusout fallback closed the popup while its input had focus.
  Both checks now use `getActiveElement` (same pattern as FloatingLinkEditorPlugin)

### Dedup
- Extracted `$unwrapRubyNode`; removes 3 copies of the ruby→TextNode unwrap logic

### Tests added
- Unit (6): element-point arrows (incl. must-not-consume at boundaries so the caret can
  leave the paragraph) and element-point backspace
- E2E: repeated Shift+Right stall guard (fails without the offset≥1 landing);
  2 shadow-DOM floating-editor tests (the focusout one fails against `0a8698d`)

### Verification
- 77/77 unit; 25/25 e2e on Chromium, Firefox, and WebKit; tsc/ESLint/Prettier clean
- Both behavioral fixes proven differentially: each new guard test fails on `0a8698d`

</details>

**Review by etrepum — APPROVED** [review]

**Comment by mayrang** [comment]
Reviewed and tested the NodeCaret refactor + shadow-aware focus commit locally — unit tests pass, tsc clean. LGTM, ready to merge whenever you are.
